### PR TITLE
[DLP] Implementation of computing K Anonymity With EntityId

### PR DESCRIPTION
**Pull request by bhavika-dhanwani**
Implementation of [computing K Anonymity With EntityId.](https://cloud.google.com/dlp/docs/visualizing_re-id_risk.md?g=0&l=119#step_1_calculate_k-anonymity_on_the_dataset)
## Checklist

- [X] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [ ] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [X] **Tests** pass:   `mvn clean verify` **required**
- [X] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [X] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  public static void calculateKAnonymityWithEntityId(
+      String projectId, String datasetId, String tableId) throws IOException {
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests. After completing all of your requests, call
+    // the "close" method on the client to safely clean up any remaining background resources.
+    FileInputStream credentialsFile =
+        new FileInputStream(
+            "/Users/bhavika.dhanwani/Downloads/bdp-2059-is-31084-c1d4f0928c82.json");
```
Comment (soumya92) [inline dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:65]: Don't we have a standard client? None of the other samples had to use credential files like this.

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      List<String> quasiIds = Arrays.asList("Age", "Mystery");
+
+      // Configure the privacy metric to compute for re-identification risk analysis.
+      List<FieldId> quasiIdFields =
+          quasiIds.stream()
+              .map(columnName -> FieldId.newBuilder().setName(columnName).build())
+              .collect(Collectors.toList());
+
+      // Specify the unique identifier in the source table for the k-anonymity analysis.
+      FieldId fieldId = FieldId.newBuilder().setName("Name").build();
```
Comment (soumya92) [inline dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:93]: `FieldId uniqueIdField`

File: dlp/snippets/src/test/java/dlp/snippets/RiskAnalysisTests.java
Comment (soumya92) [inline dlp/snippets/src/test/java/dlp/snippets/RiskAnalysisTests.java:1]: Why the change to make `DLP_SERVICE_CLIENT` static? It's better to minimise state that's shared between tests to avoid issues.

**Review by soumya92 — CHANGES_REQUESTED** [review]

File: dlp/snippets/src/test/java/dlp/snippets/RiskAnalysisTests.java
Comment (bhavika-dhanwani) [reply dlp/snippets/src/test/java/dlp/snippets/RiskAnalysisTests.java:1]: @soumya92 I received feedback in a couple of PRs to avoid initialising the DLP client repeatedly for each test case, so I have done the same changes in this PR too 
https://github.com/GoogleCloudPlatform/java-docs-samples/pull/7924/files/f05d0adb2db695e8f9e903b082b52d0e734a0a02#r1176643367

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      // Configure the privacy metric to compute for re-identification risk analysis.
+      List<FieldId> quasiIdFields =
+          quasiIds.stream()
+              .map(columnName -> FieldId.newBuilder().setName(columnName).build())
+              .collect(Collectors.toList());
+
+      // Specify the unique identifier in the source table for the k-anonymity analysis.
+      FieldId uniqueIdField = FieldId.newBuilder().setName("Name").build();
+      EntityId entityId = EntityId.newBuilder().setField(uniqueIdField).build();
+      PrivacyMetric.KAnonymityConfig kanonymityConfig =
```
Comment (Sita04) [inline dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:85]:
```suggestion
      KAnonymityConfig kanonymityConfig =
```
Include the PrivacyMetric.KAnonymityConfig in imports.

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+          PrivacyMetric.KAnonymityConfig.newBuilder()
```
Comment (Sita04) [inline dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:86]:
```suggestion
          KAnonymityConfig.newBuilder()
```

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+      // Build a request to get the completed job
+      GetDlpJobRequest getDlpJobRequest =
+          GetDlpJobRequest.newBuilder().setName(dlpJob.getName()).build();
+
+      DlpJob completedJob = null;
+      do {
+        completedJob = dlpServiceClient.getDlpJob(getDlpJobRequest);
+        Thread.sleep(10000);
+      } while (completedJob.getState() != DlpJob.JobState.DONE);
```
[the remark this answers is not in the document]
Comment (Sita04) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:135]: Related comment here: https://github.com/GoogleCloudPlatform/java-docs-samples/pull/8013#discussion_r1201841469

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+      PrivacyMetric privacyMetric =
+          PrivacyMetric.newBuilder().setKAnonymityConfig(kanonymityConfig).build();
+
+      // Specify the bigquery table to store the findings.
+      BigQueryTable outputbigQueryTable =
+          BigQueryTable.newBuilder()
+              .setProjectId(projectId)
+              .setDatasetId(datasetId)
+              .setTableId("test_results")
```
[the remark this answers is not in the document]
Comment (bhavika-dhanwani) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:99]: The "test_results" table will be created if it doesn't exist in the BigQuery dataset, so it's not parameterized.

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+import com.google.privacy.dlp.v2.PrivacyMetric;
+import com.google.privacy.dlp.v2.RiskAnalysisJobConfig;
+import com.google.privacy.dlp.v2.Value;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("checkstyle:AbbreviationAsWordInName")
```
[the remark this answers is not in the document]
Comment (bhavika-dhanwani) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:44]: While executing `mvn -P lint checkstyle:check`, one linting check was failing since the class name `RiskAnalysisKAnonymityWithEntityId` contains consecutive capital letters, so have suppressed this warning

**Review by soumya92 — APPROVED** [review]

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhavika-dhanwani) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:136]: Using `.getDlpJobCallable()` returns the result before the job is completed, so do{ } while loop seems the only approach for now

[the remark this answers is not in the document]
Comment (Sita04) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:136]: Okay, sounds good.

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sita04) [reply dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:100]: Could you mention this in a comment?

**Review by Sita04 — APPROVED** [review]
LGTM. Just one minor comment.

**Comment by bhavika-dhanwani** [comment]
> LGTM. Just one minor comment.

Done

File: dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    calculateKAnonymityWithEntityId(projectId, datasetId, tableId);
+  }
+
+  // Uses the Data Loss Prevention API to compute the k-anonymity of a column set in a Google
+  // BigQuery table.
+  public static void calculateKAnonymityWithEntityId(
+      String projectId, String datasetId, String tableId) throws IOException, InterruptedException {
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests. After completing all of your requests, call
+    // the "close" method on the client to safely clean up any remaining background resources.
```
Comment (anguillanneuf) [inline dlp/snippets/src/main/java/dlp/snippets/RiskAnalysisKAnonymityWithEntityId.java:65]: This comment is outdated because `close()` is not called.
